include/ibamr/AdvDiffComplexFluidConvectiveOperator.h

```diff
@@ -0,0 +1,173 @@
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef included_AdvDiffOldroydBConvectiveOperator
+#define included_AdvDiffOldroydBConvectiveOperator
+/////////////////////////////// INCLUDES /////////////////////////////////////
+#include "IBAMR_config.h"
+
+#include "ibamr/AdvDiffCenteredConvectiveOperator.h"
+#include "ibamr/AdvDiffPPMConvectiveOperator.h"
+#include "ibamr/AdvDiffPhysicalBoundaryUtilities.h"
+#include "ibamr/AdvDiffWavePropConvectiveOperator.h"
+#include "ibamr/CFRelaxationOperator.h"
+#include "ibamr/ConvectiveOperator.h"
+#include "ibamr/INSHierarchyIntegrator.h"
+#include "ibamr/StaggeredStokesPhysicalBoundaryHelper.h"
+#include "ibamr/ibamr_enums.h"
+#include "ibamr/ibamr_utilities.h"
+#include "ibamr/namespaces.h" // IWYU pragma: keep
+
+#include "ibtk/CartExtrapPhysBdryOp.h"
+#include "ibtk/HierarchyGhostCellInterpolation.h"
+#include "ibtk/ibtk_utilities.h"
+
+#include "Box.h"
+#include "CartesianGridGeometry.h"
+#include "CartesianPatchGeometry.h"
+#include "CellData.h"
+#include "CellVariable.h"
+#include "FaceData.h"
+#include "FaceVariable.h"
+#include "Index.h"
+#include "IntVector.h"
+#include "Patch.h"
+#include "PatchHierarchy.h"
+#include "PatchLevel.h"
+#include "SAMRAIVectorReal.h"
+#include "Variable.h"
+#include "VariableContext.h"
+#include "VariableDatabase.h"
+#include "tbox/Database.h"
+#include "tbox/Pointer.h"
+#include "tbox/Utilities.h"
+
+#include <string>
+#include <vector>
+
+namespace IBAMR
+{
+enum CFConvectiveOperatorType
+{
+    CENTERED,
+    WAVE_PROP,
+    CF_PPM,
+    UNKNOWN = -1
+};
+template <>
+inline std::string
+enum_to_string<CFConvectiveOperatorType>(CFConvectiveOperatorType val)
+{
+    if (val == CENTERED) return "CENTERED";
+    if (val == WAVE_PROP) return "WAVE_PROP";
+    if (val == CF_PPM) return "PPM";
+    return "UNKNOWN";
+}
+template <>
+inline CFConvectiveOperatorType
+string_to_enum<CFConvectiveOperatorType>(const std::string& str)
+{
+    if (strcasecmp(str.c_str(), "CENTERED") == 0) return CENTERED;
+    if (strcasecmp(str.c_str(), "WAVE_PROP") == 0) return WAVE_PROP;
+    if (strcasecmp(str.c_str(), "PPM") == 0) return CF_PPM;
+    return UNKNOWN;
+}
+/*!
+ * \brief Class AdvDiffComplexFluidConvectiveOperator is a concrete ConvectiveOperator that implements a convective
```

Comment:
This is really an UCD for the conformation tensor or extra stress, right? I wonder if it would be both clearer and shorter to name this something like `CCUpperConvectiveOperator` (`CC` = cell-centered).

Reply:
Pretty much. The way it's written now, a relaxation function has to be registered.

The relaxation function could just be registered as a source term for the advection diffusion integrator, which would leave this as an UCD operator. That would make it more general. I'll do that, and move the UCD operator into the `adv_diff` folder.

Reply:
Actually, that's not completely true. This advects the conformation tensor _or_ the square root or logarithm of the tensor. So it's not just a UCD operator.